src/main/java/characterpicker/infrastructure/CharacterParser.java

```diff
@@ -3,26 +3,55 @@
 import characterpicker.domain.Character;
 import org.apache.commons.lang.StringUtils;
 
+import java.util.ArrayList;
+import java.util.List;
+import java.util.Locale;
+
 public class CharacterParser {
-    private CharacterParser(){
+    private CharacterParser() {
         throw new IllegalStateException("Utility class");
     }
 
     static Character parseCharacter(String unparsedCharacter) {
 
-        String[] characterProperties = unparsedCharacter.split(",");
+        if (unparsedCharacter.endsWith(";")){
+            throw new IllegalArgumentException("Completion required. If you don't have one type 'none'");
+        }
+
+        String[] unparsedProperties = unparsedCharacter.split(";");
+        String character = unparsedProperties[0];
+        String completion = unparsedProperties[1];
+
+        String[] characterProperties = character.split(",");
         if (characterProperties.length < 2) {
             throw new IllegalArgumentException("Name and weight required");
         }
         String name = characterProperties[0];
         String unparsedWeight = characterProperties[1];
+
         if (!StringUtils.isNumeric(unparsedWeight)) {
             throw new IllegalArgumentException("Provided weight is not a number");
         }
 
-
-        int weight = Integer.parseInt(unparsedWeight);
+        int weight = Integer.parseInt(unparsedWeight) + parseCompletion(completion);
 
         return new Character(name, weight);
     }
+    //TO DO change list name
+
+    private static Integer parseCompletion(String completion) {
+        String[] completionProperties = completion.split(",");
+        List<String> bosses = new ArrayList<>();
+
+        int length = completionProperties.length;
+        int i = 0;
+
+        while (length > 0) {
+            bosses.add(completionProperties[i].toLowerCase(Locale.ROOT));
+            length--;
+            i++;
```

Comment:
Po pierwsze:

Nie musisz redukować `length` i zwiększać `i` wystarczy, że w tym while uzależnisz `length` od `i` `while(length > i)`

Po drugie:

Do iteracji po określonej ilości elementów, bardziej nadaje się do tego pętla `for` - jak musisz odnosić się do indexu, albo pętla `foreach` jak index Cię nie obchodzi tylko zależy Ci na tym, aby przejść przez wszystkie elementy (czyli tak jak tutaj)

Po trzecie można to w miarę clean zrobić streamem.

```java
        List<String> bosses = Arrays.stream(completion.split(","))
                                    .map(bossMark -> bossMark.toLowerCase(Locale.ROOT))
                                    .collect(Collectors.toList());
```